url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/VmSchedulerTimeShared.java

```diff
@@ -0,0 +1,250 @@
+			// each virtual PE of a VM must require not more than the capacity of a physical PE
+			if (mips > peMips) {
+				return false;
+			}
+			totalRequestedMips += mips;
+		}
+
+		// This scheduler does not allow over-subscription
+		if (getAvailableMips() < totalRequestedMips) {
+			return false;
+		}
+
+		getMipsMapRequested().put(vmUid, mipsShareRequested);
+		setPesInUse(getPesInUse() + mipsShareRequested.size());
+
+		if (getVmsMigratingIn().contains(vmUid)) {
+			// the destination host only experience 10% of the migrating VM's MIPS
+			totalRequestedMips *= 0.1;
+		}
+
+		List<Double> mipsShareAllocated = new ArrayList<Double>();
+		for (Double mipsRequested : mipsShareRequested) {
+			if (getVmsMigratingOut().contains(vmUid)) {
+				// performance degradation due to migration = 10% MIPS
+				mipsRequested *= 0.9;
+			} else if (getVmsMigratingIn().contains(vmUid)) {
+				// the destination host only experience 10% of the migrating VM's MIPS
+				mipsRequested *= 0.1;
+			}
+			mipsShareAllocated.add(mipsRequested);
+		}
+
+		getMipsMap().put(vmUid, mipsShareAllocated);
+		setAvailableMips(getAvailableMips() - totalRequestedMips);
+
+		return true;
+	}
+
+	/**
+	 * Update allocation of VMs on PEs.
+	 */
+	protected void updatePeProvisioning() {
+                //Log.printLine("dans updatePeProvisioning");
+		getPeMap().clear();
+		for (Pe pe : getPeList()) {
+			pe.getPeProvisioner().deallocateMipsForAllVms();
+		}
+
+		Iterator<Pe> peIterator = getPeList().iterator();
+		Pe pe = peIterator.next();
+		PeProvisioner peProvisioner = pe.getPeProvisioner();
+		double availableMips = peProvisioner.getAvailableMips();
+                //Log.printLine("availableMips = " + availableMips);
+                
+		for (Map.Entry<String, List<Double>> entry : getMipsMap().entrySet()) {
+                    
+			String vmUid = entry.getKey();
+			getPeMap().put(vmUid, new LinkedList<Pe>());
+
+			for (double mips : entry.getValue()) {
+                    //        Log.printLine("mips : " + mips + " // available mips : " + availableMips);
+				while (mips >= 0.1) {
+					if (availableMips >= mips) {
+						peProvisioner.allocateMipsForVm(vmUid, mips);
+						getPeMap().get(vmUid).add(pe);
+						availableMips -= mips;
+						break;
+					} else {
+						peProvisioner.allocateMipsForVm(vmUid, availableMips);
+						getPeMap().get(vmUid).add(pe);
+						mips -= availableMips;
+						if (mips <= 0.1) {
+							break;
+						}
+						if (!peIterator.hasNext()) {
+							Log.printLine("There is no enough MIPS (" + mips + ") to accommodate VM " + vmUid);
+							// System.exit(0);
+						}
+						pe = peIterator.next();
+						peProvisioner = pe.getPeProvisioner();
+						availableMips = peProvisioner.getAvailableMips();
+					}
+				}
+			}
+		}
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see cloudsim.VmScheduler#deallocatePesForVm(cloudsim.Vm)
+	 */
+	@Override
+	public void deallocatePesForVm(Vm vm) {
+		getMipsMapRequested().remove(vm.getUid());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;vm&#39; (index: 0 | type: Vm). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other methods validating the parameter: &#39;allocatePesForVm&#39;. The same parameter type is validated here for example: url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/Host.java:164. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).